modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/QueryRegistryImpl.java

```diff
@@ -21,50 +21,85 @@
 import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
-
-import org.apache.ignite.IgniteLogger;
+import java.util.stream.Collectors;
+import org.apache.calcite.util.Pair;
+import org.apache.ignite.cache.query.QueryCancelledException;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.cache.query.GridCacheQueryType;
+import org.apache.ignite.internal.processors.query.GridQueryCancel;
+import org.apache.ignite.internal.processors.query.GridRunningQueryManager;
 import org.apache.ignite.internal.processors.query.RunningQuery;
+import org.apache.ignite.internal.processors.query.calcite.exec.ddl.RunningQueryManagerWrapper;
+import org.apache.ignite.internal.processors.query.calcite.util.AbstractService;
 import org.apache.ignite.internal.util.IgniteUtils;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * Registry of the running queries.
  */
-public class QueryRegistryImpl implements QueryRegistry {
+public class QueryRegistryImpl extends AbstractService implements QueryRegistry {
     /** */
-    private final ConcurrentMap<UUID, RunningQuery> runningQrys = new ConcurrentHashMap<>();
+    private final ConcurrentMap<UUID, Pair<Long, RunningQuery>> runningQrys = new ConcurrentHashMap<>();
 
     /** */
-    private final IgniteLogger log;
+    private final GridRunningQueryManager runningQryMgr;
 
     /** */
-    public QueryRegistryImpl(IgniteLogger log) {
-        this.log = log;
+    public QueryRegistryImpl(GridKernalContext ctx) {
+        super(ctx);
+
+        runningQryMgr = new RunningQueryManagerWrapper(ctx);
     }
 
     /** {@inheritDoc} */
-    @Override public RunningQuery register(RunningQuery qry) {
-        return runningQrys.computeIfAbsent(qry.id(), k -> qry);
+    @Override public RunningQuery register(String sql, String schema, RunningQuery qry) {
+        return runningQrys.computeIfAbsent(qry.id(), k -> {
+            String nodeId;
+            if (qry.initiatorNodeId() != null)
+                nodeId = qry.initiatorNodeId().toString();
+            else
+                nodeId = kctx.discovery().localNode().id().toString();
+
+            Long locId = runningQryMgr.register(sql, GridCacheQueryType.SQL_FIELDS, schema, false,
+                createCancelToken(qry), nodeId);
+
+            return Pair.of(locId, qry);
```

Comment:
Perhaps we can store `locId` in query, instead of creating Pair. WDYT?

Reply:
I thought that it is better for calcite-3